### PR TITLE
bestpractices: add crash pre-requisites

### PR DESCRIPTION
**Pull request by gireeshpunathil**
This covers the preparation of crash diagnostics
Few caveats:
 -  I am not good at markup, so please advise on the structure
 -  As this will form a baseline for the rest of the docs, a thorough review is requested: both in the format as well as the content.
thanks!

Refs: https://github.com/nodejs/diagnostics/issues/254

### CODE REVIEW

File: documentation/crash/crash_setup.md [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+ - in certain deployment scenarios (such as Cloud) those
+   operate in a state-less mode [ no persistent disk]
+
+The net effect of these constraints is that your production systems
+need to be manually `prepared` in advance to enable crash dianostic
+data generation on the first failure itself, without loosing vital data.
+The rest of the document illustrates this preparation steps.
+ 
+## Available  disk space
+Ensure that there is enough disk space available for the core file
```
Comment (mhdawson) [inline documentation/crash/crash_setup.md]:
```suggestion
Ensure that there is enough disk space available for 1 or more core files
```

File: documentation/crash/crash_setup.md [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+to be written:
+
+ - Maximum of 4GB for a 32 bit process.
+ - Much larger  for  64 bit process (common case). To know the precise
+   requirement, measure the peak-load memory usage of your application.
+   Add a 10% to that to accommodate core metadata. If you are using
```
Comment (mhdawson) [inline documentation/crash/crash_setup.md]:
```suggestion
Add 10% to that to accommodate core metadata. If you are using
 ```

File: documentation/crash/crash_setup.md [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+   common monitoring tools, one of the graph should reveal the peak
```
Comment (mhdawson) [inline documentation/crash/crash_setup.md]:
```suggestion
   common monitoring tools, one of the graphs should reveal the peak
```

File: documentation/crash/crash_setup.md [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+and can severely hamper the ability to diagnose the problem.
+
+Figure out how much free space is available in the file system:
+`df -k` can be used invariably across UNIX platforms.
+In Windows, Windows explorer when pointed to a disk partition,
+provides a view of the available space in that partition.
+
+## Core file location and name
+
+By default, core file is generated on a crash event, and is
```
Comment (mhdawson) [inline documentation/crash/crash_setup.md]:
```suggestion
By default, a core file is generated when a crash occurs, and is
```

File: documentation/crash/crash_setup.md [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+In freebsd and Darwin, use `gcore [-s] [executable] pid`
+In Windows, you can use `Task manager` window, right click on the
+node process and select `create dump` option.
+
+Special note on Ubuntu systems with `Yama hardened kernel`
+
+Yama security policy inhibits a second process from collecting dump,
+practically rendering `gcore` unusable.
+
+`setcap cap_sys_ptrace=+ep `which gdb``
```
Comment (mhdawson) [inline documentation/crash/crash_setup.md]: It's not clear to me if this is to address the problem of gcore not working?

**Comment by mhdawson** [comment]
Pushed a few commits to do minor fixup and some suggested changes as well.

A more structural suggestion is that we should organize it into the following format:

## Introduction

What you have for the introduction with additional paragraphs for

 +  mentioning the key artifacts for exploring crashes (core dumps, diagnostic report?)

## Common issues

Cover common issues in generating the key artifacts.

* Disk space
* ulimits

## Recommended best practice

This section provides specific recommendations for how to configure your systems in advance to be ready in order to be ready to investigate crashes.

### Ensuring enough disk space

your existing content on how to size.  Maybe talk about cleanup/management of cores to avoid space problems here instead of later on?

### Configuring to ensure core generation

* ulmits (what you have + would be good to have a recommendation on how to check/configure the har limits)
* specific recommendation for core file naming/patterns and how to set them
* recommendations for cleanup/management of cores that are generated (otherwise can result in out of space for application)


## Additional information 

* Manual Dump generation


Let me know what you think about this structure/organization.

File: documentation/crash/crash_setup.md (hunk 5 above)
Comment (gireeshpunathil) [reply documentation/crash/crash_setup.md]: @mhdawson - addressed, added a line that explains what we are doing.

**Comment by gireeshpunathil** [comment]
@mhdawson - I have addressed all of your review comments. PTAL, thanks!

**Comment by gireeshpunathil** [comment]
thanks @mhdawson . Can I get one more review please!

File: documentation/crash/crash_setup.md [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+and staging systems in many aspects:
+
+ - they are isolated from public internet
+ - they are not loaded with development and debug tools
+ - they are configured with the most robust and secure
+   configurations possible at the OS level
+ - in certain deployment scenarios (such as Cloud) those
+   operate in a head-less mode [ no ssh ]
+ - in certain deployment scenarios (such as Cloud) those
+   operate in a state-less mode [ no persistent disk]
```
Comment (hekike) [inline documentation/crash/crash_setup.md]: maybe you could mention the performance constraint, that you cannot usually use tools that slow down the process significantly

File: documentation/crash/crash_setup.md [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+
+The net effect of these constraints is that your production systems
+need to be manually `prepared` in advance to enable crash dianostic
+data generation on the first failure itself, without loosing vital data.
+The rest of the document illustrates this preparation steps.
+ 
+The key artifacts for exploring Node.js application crashes are:
```
Comment (hekike) [inline documentation/crash/crash_setup.md]: maybe adding `in production`

File: documentation/crash/crash_setup.md [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+
+Workaround this by enabling `ptrace` capability to gdb.
+Execute the below as root:
+
+`setcap cap_sys_ptrace=+ep `which gdb``
+
+
+These steps make sure that when your Node.js application crashes in
+production a valid, full core dump is generated at a known location that
+can be loaded into debuggers that understand Node.js internsls, and
```
[the remark this answers is not in the document]
Comment (gireeshpunathil) [reply documentation/crash/crash_setup.md]: done, thanks!

**Comment by gireeshpunathil** [comment]
> Additionally, you could call out that the user will need the exact same node binary to inspect your core dump. Or they need to bundle it into the core dump.

@hekike - can I add that statement in my second part? as here we are talking about the production system, and this one is for the `dev host` where we will be performing crash analysis, it makes sense to talk about that when we mention the debugging basics? PLMK.

**Comment by hekike** [comment]
@gireeshpunathil makes sense!
Thanks for the PR!

**Comment by gireeshpunathil** [comment]
landed as 38a577a82beef0176e5a7147f40c82a9e87402b8

**Pull request closed without merging**